File: rust/crates/mokume-quant/src/maxlfq.rs
//! Maximal peptide ratio extraction: Cox et al. (2014), Eq. 3 and Fig. 2.
//!
//! Input is a single protein's positive, linear peptide-species intensities.
//! Species must distinguish sequence, modification and charge. Normalization
//! (including the paper's separate delayed-normalization stage) is upstream.

use std::collections::{BTreeMap, HashMap};

use mokume_core::{MokumeError, Result, SampleId};

use crate::{median, sample_order, PeptideMeasurement};

/// The minimum number of shared species supporting a sample-pair ratio.
pub const DEFAULT_MAXLFQ_MIN_RATIO_COUNT: usize = 2;

#[derive(Debug, Clone)]
pub struct MaxLfqResult {
    /// Linear intensities in the requested sample order; zero means unquantified.
    pub intensities: Vec<(SampleId, f64)>,
    /// Independently solved sample components. Ratios across components are not
    /// identifiable from shared peptides; their offsets use observed totals.
    pub components: Vec<Vec<SampleId>>,
    /// Valid sample pairs assigned a positive large-ratio stabilization weight.
    pub stabilized_pairs: usize,
}

pub fn max_lfq(measurements: &[PeptideMeasurement]) -> Result<Vec<(SampleId, f64)>> {
    max_lfq_with_samples(measurements, &sample_order(measurements))
}

pub fn max_lfq_with_samples(
    measurements: &[PeptideMeasurement],
    samples: &[SampleId],
) -> Result<Vec<(SampleId, f64)>> {
    Ok(solve_max_lfq(measurements, samples, DEFAULT_MAXLFQ_MIN_RATIO_COUNT)?.intensities)
}

/// Minimize the unweighted squared residuals of all valid pairwise log ratios.
/// Duplicate species/sample observations are summed before taking logarithms.
/// Missing, nonpositive and nonfinite observations do not support a ratio.
/// A sample without a valid edge stays zero (Cox et al., Fig. 2D).
pub fn solve_max_lfq(
    measurements: &[PeptideMeasurement],
    samples: &[SampleId],
    min_ratio_count: usize,
) -> Result<MaxLfqResult> {
    solve_max_lfq_with_stabilization(measurements, samples, min_ratio_count, false)
}

/// Apply Cox et al. Eq. 5 to low-overlap ratios when `stabilize` is enabled.
/// Shared-species requirements and disconnected-component handling are unchanged.
pub fn solve_max_lfq_with_stabilization(
    measurements: &[PeptideMeasurement],
    samples: &[SampleId],
    min_ratio_count: usize,
    stabilize: bool,
) -> Result<MaxLfqResult> {
    if min_ratio_count == 0 {
        return Err(invalid("minimum ratio count must be positive"));
    }
    let sample_index: HashMap<_, _> = samples.iter().enumerate().map(|(i, s)| (*s, i)).collect();
    if sample_index.len() != samples.len() {
        return Err(invalid("sample identifiers must be unique"));
    }
    let rows = intensity_rows(measurements, &sample_index)?;
    let (edges, stabilized_pairs) =
        pairwise_ratios(&rows, samples.len(), min_ratio_count, stabilize);
    let components = connected_components(&edges, samples.len());
    let mut log_profile = vec![f64::NEG_INFINITY; samples.len()];
    for component in &components {
        let relative = solve_component(component, &edges, samples.len())?;
        let total = log_total(&rows, component.iter().copied());
        let shift = total - log_sum_exp2(&relative);
        for (&sample, value) in component.iter().zip(relative) {
            log_profile[sample] = value + shift;
        }
    }
    // Eq. 3 fixes ratios, not the absolute scale. Rescale the entire supported
    // profile to the original summed intensity, including isolated observations.
    if !components.is_empty() {
        let shift = log_total(&rows, 0..samples.len()) - log_sum_exp2(&log_profile);
        for value in &mut log_profile {
            *value += shift;
        }
    }
    Ok(MaxLfqResult {
        intensities: linear_intensities(samples, log_profile)?,
        stabilized_pairs,
        components: components
            .into_iter()
            .map(|c| c.into_iter().map(|i| samples[i]).collect())
            .collect(),
    })
}

fn linear_intensities(samples: &[SampleId], log_profile: Vec<f64>) -> Result<Vec<(SampleId, f64)>> {
    samples
        .iter()
        .zip(log_profile)
        .map(|(&sample, log_value)| {
            let value = log_value.exp2();
            if !value.is_finite() || (log_value.is_finite() && value == 0.0) {
                return Err(invalid("protein intensity is outside the finite f64 range"));
            }
            Ok((sample, value))
        })
        .collect()
}

type RatioEdge = (usize, usize, f64);

fn intensity_rows(
    measurements: &[PeptideMeasurement],
    sample_index: &HashMap<SampleId, usize>,
) -> Result<Vec<Vec<f64>>> {
    let mut ordered = measurements
        .iter()
        .filter(|m| {
            m.intensity.is_finite() && m.intensity > 0.0 && sample_index.contains_key(&m.sample)
        })
        .collect::<Vec<_>>();
    ordered.sort_by(|a, b| {
        (a.peptide.get(), a.sample.get())
            .cmp(&(b.peptide.get(), b.sample.get()))
            .then_with(|| a.intensity.total_cmp(&b.intensity))
    });
    let mut rows = BTreeMap::<u32, Vec<f64>>::new();
    for measurement in ordered {
        let values = rows
            .entry(measurement.peptide.get())
            .or_insert_with(|| vec![0.0; sample_index.len()]);
        let value = &mut values[sample_index[&measurement.sample]];
        *value += measurement.intensity;
        if !value.is_finite() {
            return Err(invalid(
                "summed species intensity is outside the finite f64 range",
            ));
        }
    }
    Ok(rows.into_values().collect())
}

fn pairwise_ratios(
    rows: &[Vec<f64>],
    n: usize,
    min_count: usize,
    stabilize: bool,
) -> (Vec<RatioEdge>, usize) {
    let logs = rows
        .iter()
        .map(|row| row.iter().map(|x| x.log2()).collect::<Vec<_>>())
        .collect::<Vec<_>>();
    let mut edges = Vec::new();
    let mut stabilized_pairs = 0;
    let mut ratios = Vec::with_capacity(rows.len());
    for i in 0..n {
        for j in i + 1..n {
            ratios.clear();
            ratios.extend(logs.iter().filter_map(|row| {
                (row[i].is_finite() && row[j].is_finite()).then_some(row[j] - row[i])
            }));
            if ratios.len() >= min_count {
                if let Some(ratio) = median(&mut ratios) {
                    let (ratio, applied) = if stabilize {
                        stabilized_ratio(rows, (i, j), ratios.len(), ratio)
                    } else {
                        (ratio, false)
                    };
                    stabilized_pairs += usize::from(applied);
                    edges.push((i, j, ratio));
                }
            }
        }
    }
    (edges, stabilized_pairs)
}

fn stabilized_ratio(
    rows: &[Vec<f64>],
    (i, j): (usize, usize),
    shared: usize,
    median_ratio: f64,
) -> (f64, bool) {
    let count_i = rows.iter().filter(|row| row[i] > 0.0).count();
    let count_j = rows.iter().filter(|row| row[j] > 0.0).count();
    let x = count_i.max(count_j) as f64 / shared as f64;
    let weight = ((x - 2.5) / 2.5).clamp(0.0, 1.0);
    if weight == 0.0 {
        return (median_ratio, false);
    }
    let sum_ratio = log_total(rows, std::iter::once(j)) - log_total(rows, std::iter::once(i));
    ((1.0 - weight) * median_ratio + weight * sum_ratio, true)
}

fn connected_components(edges: &[RatioEdge], n: usize) -> Vec<Vec<usize>> {
    let mut neighbors = vec![Vec::new(); n];
    for &(i, j, _) in edges {
        neighbors[i].push(j);
        neighbors[j].push(i);
    }
    let mut seen = vec![false; n];
    let mut components = Vec::new();
    for start in 0..n {
        if seen[start] || neighbors[start].is_empty() {
            continue;
        }
        let mut pending = vec![start];
        let mut component = Vec::new();
        seen[start] = true;
        while let Some(i) = pending.pop() {
            component.push(i);
            for &j in &neighbors[i] {
                if !seen[j] {
                    seen[j] = true;
                    pending.push(j);
                }
            }
        }
        component.sort_unstable();
        components.push(component);
    }
    components
}

fn solve_component(component: &[usize], edges: &[RatioEdge], samples: usize) -> Result<Vec<f64>> {
    let n = component.len() - 1;
    let mut indices = vec![None; samples];
    for (i, &sample) in component.iter().enumerate() {
        indices[sample] = Some(i);
    }
    // Fix the first log intensity to zero. The reduced graph Laplacian is
    // positive definite for a connected component; no ridge or fallback is used.
    let mut matrix = vec![vec![0.0; n]; n];
    let mut rhs = vec![0.0; n];
    for &(left, right, ratio) in edges {
        if let (Some(i), Some(j)) = (indices[left], indices[right]) {
            if i > 0 {
                matrix[i - 1][i - 1] += 1.0;
                rhs[i - 1] -= ratio;
            }
            if j > 0 {
                matrix[j - 1][j - 1] += 1.0;
                rhs[j - 1] += ratio;
            }
            if i > 0 && j > 0 {
                matrix[i - 1][j - 1] -= 1.0;
                matrix[j - 1][i - 1] -= 1.0;
            }
        }
    }
    let solution = cholesky_solve(matrix, rhs)?;
    Ok(std::iter::once(0.0).chain(solution).collect())
}

fn cholesky_solve(mut matrix: Vec<Vec<f64>>, mut rhs: Vec<f64>) -> Result<Vec<f64>> {
    let n = rhs.len();
    for i in 0..n {
        for j in 0..=i {
            let product: f64 = matrix[i][..j]
                .iter()
                .zip(&matrix[j][..j])
                .map(|(a, b)| a * b)
                .sum();
            let value = matrix[i][j] - product;
            matrix[i][j] = if i == j {
                if value <= 0.0 || !value.is_finite() {
                    return Err(invalid("connected-component linear solve failed"));
                }
                value.sqrt()
            } else {
                value / matrix[j][j]
            };
        }
        let product: f64 = matrix[i][..i]
            .iter()
            .zip(&rhs[..i])
            .map(|(a, b)| a * b)
            .sum();
        rhs[i] = (rhs[i] - product) / matrix[i][i];
    }
    for i in (0..n).rev() {
        let product: f64 = matrix[i + 1..]
            .iter()
            .zip(&rhs[i + 1..])
            .map(|(row, value)| row[i] * value)
            .sum();
        rhs[i] = (rhs[i] - product) / matrix[i][i];
    }
    Ok(rhs)
}

fn log_sum_exp2(values: &[f64]) -> f64 {
    let max = values.iter().copied().fold(f64::NEG_INFINITY, f64::max);
    max + values.iter().map(|x| (x - max).exp2()).sum::<f64>().log2()
}

fn log_total(rows: &[Vec<f64>], columns: impl Iterator<Item = usize>) -> f64 {
    let columns = columns.collect::<Vec<_>>();
    let mut logs = rows
        .iter()
        .flat_map(|row| columns.iter().map(|&i| row[i].log2()))
        .collect::<Vec<_>>();
    // Stable under peptide and sample permutations, including rounding of sums.
    logs.sort_by(f64::total_cmp);
    log_sum_exp2(&logs)
}

fn invalid(message: &str) -> MokumeError {
    MokumeError::InvalidInput {
        message: format!("MaxLFQ: {message}"),
    }
}

#[cfg(test)]
mod tests;

File: python/mokume/quantification/maxlfq.py
"""Maximal peptide ratio extraction from Cox et al. (2014), Eq. 3/Fig. 2.

This module reconstructs protein profiles from pairwise peptide-species ratios.
Normalization is an explicit upstream step; fraction-aware delayed normalization
is not performed here. Optional large-ratio stabilization follows Eq. 5.
Source: https://doi.org/10.1074/mcp.M113.031591
"""

import math
from dataclasses import dataclass, field
from typing import ClassVar, Optional

import numpy as np
import pandas as pd
from joblib import Parallel, delayed

from mokume.core.constants import (
    NORM_INTENSITY,
    PEPTIDE_CANONICAL,
    PROTEIN_NAME,
    SAMPLE_ID,
)
from mokume.core.logger import get_logger
from mokume.core.registry import PluginRegistry
from mokume.quantification.base import ProteinQuantificationMethod

logger = get_logger("mokume.quantification.maxlfq")


def _identifier_sort_key(value: object) -> tuple[str, str, object]:
    value_type = type(value)
    return value_type.__module__, value_type.__qualname__, value


def _ratio_system(log_matrix: np.ndarray, min_count: int, stabilize: bool):
    """Build Eq. 3's unweighted graph Laplacian and right-hand side."""
    n_samples = log_matrix.shape[1]
    laplacian = np.zeros((n_samples, n_samples))
    rhs = np.zeros(n_samples)
    stabilized_pairs = 0
    for i in range(n_samples):
        for j in range(i + 1, n_samples):
            valid = np.isfinite(log_matrix[:, i]) & np.isfinite(log_matrix[:, j])
            if np.count_nonzero(valid) < min_count:
                continue
            ratio = np.median(log_matrix[valid, j] - log_matrix[valid, i])
            if stabilize:
                ratio, applied = _stabilized_ratio(log_matrix[:, [i, j]], valid, ratio)
                stabilized_pairs += int(applied)
            laplacian[i, i] += 1
            laplacian[j, j] += 1
            laplacian[i, j] = laplacian[j, i] = -1
            rhs[i] -= ratio
            rhs[j] += ratio
    return laplacian, rhs, stabilized_pairs


def _stabilized_ratio(log_pair, shared_mask, median_ratio):
    """Apply Cox Eq. 5 in log2 space after the minimum shared count is met."""
    counts = np.isfinite(log_pair).sum(axis=0)
    x = counts.max() / np.count_nonzero(shared_mask)
    weight = np.clip((x - 2.5) / 2.5, 0.0, 1.0)
    if weight == 0:
        return median_ratio, False
    sum_ratio = _log_total(log_pair[:, 1]) - _log_total(log_pair[:, 0])
    return (1 - weight) * median_ratio + weight * sum_ratio, True


def _components(laplacian: np.ndarray):
    """Find connected sample components; isolated samples remain unquantified."""
    seen = set()
    for start in range(len(laplacian)):
        if start in seen or laplacian[start, start] == 0:
            continue
        pending = [start]
        seen.add(start)
        component = []
        while pending:
            i = pending.pop()
            component.append(i)
            for j in np.flatnonzero(laplacian[i] < 0):
                if j not in seen:
                    seen.add(j)
                    pending.append(j)
        yield sorted(component)


def _log_total(log_values: np.ndarray) -> float:
    values = np.asarray(log_values).ravel()
    values = values[np.isfinite(values)]
    maximum = values.max()
    return float(maximum + math.log2(math.fsum(np.exp2(values - maximum))))


def _maxlfq_solve_protein(
    peptide_matrix: np.ndarray,
    peptide_ids: Optional[np.ndarray] = None,
    min_ratio_count: int = 2,
    stabilize: bool = False,
) -> np.ndarray:
    """Solve a positive linear species-by-sample matrix using Cox et al. Eq. 3.

    Each valid sample-pair edge has equal weight. Medians are computed in log2
    space, including the midpoint convention for an even number of ratios.
    Samples without a valid edge return zero (Fig. 2D). Disconnected components
    are independently anchored to their observed intensity totals, then the
    whole supported profile is rescaled to the total input intensity. No ratio
    between disconnected components is identifiable from shared species.

    ``peptide_ids`` is accepted for compatibility; no reference peptide is used.
    """
    _ = peptide_ids
    return _solve_protein(peptide_matrix, min_ratio_count, stabilize)[0]


def _solve_protein(peptide_matrix, min_ratio_count, stabilize):
    """Return protein quantities and the number of stabilized sample pairs."""
    if not isinstance(stabilize, bool):
        raise ValueError("MaxLFQ stabilize must be a boolean")
    if (
        isinstance(min_ratio_count, bool)
        or not isinstance(min_ratio_count, (int, np.integer))
        or min_ratio_count < 1
    ):
        raise ValueError("MaxLFQ minimum ratio count must be a positive integer")
    matrix = np.asarray(peptide_matrix, dtype=float)
    if matrix.ndim != 2:
        raise ValueError("MaxLFQ input must be a species-by-sample matrix")
    log_matrix = np.full(matrix.shape, np.nan)
    valid = np.isfinite(matrix) & (matrix > 0)
    np.log2(matrix, out=log_matrix, where=valid)
    laplacian, rhs, stabilized_pairs = _ratio_system(
        log_matrix, min_ratio_count, stabilize
    )
    log_profile = np.full(matrix.shape[1], -np.inf)
    components = list(_components(laplacian))
    for component in components:
        free = component[1:]
        relative = np.r_[0.0, np.linalg.solve(laplacian[np.ix_(free, free)], rhs[free])]
        log_profile[component] = (
            relative + _log_total(log_matrix[:, component]) - _log_total(relative)
        )
    if components:
        log_profile += _log_total(log_matrix) - _log_total(log_profile)
    if len(components) > 1:
        logger.warning(
            "MaxLFQ has %d disconnected sample components; "
            "between-component ratios are not identifiable",
            len(components),
        )
    with np.errstate(over="ignore", under="ignore"):
        result = np.exp2(log_profile)
    if np.any(~np.isfinite(result)) or np.any(np.isfinite(log_profile) & (result == 0)):
        raise ValueError("MaxLFQ protein intensity is outside the finite float64 range")
    return result, stabilized_pairs


def _process_protein(
    protein: str,
    protein_data: pd.DataFrame,
    columns: tuple[str, str, str],
    samples: list,
    ratio_options: dict,
) -> tuple[list, int]:
    """Sum duplicate species observations, then solve a single protein."""
    peptide_column, intensity_column, sample_column = columns
    peptides = sorted(protein_data[peptide_column].unique(), key=_identifier_sort_key)
    pivot = protein_data.pivot_table(
        index=peptide_column,
        columns=sample_column,
        values=intensity_column,
        aggfunc=math.fsum,
        observed=True,
    ).reindex(index=peptides, columns=samples)
    intensities, stabilized_pairs = _solve_protein(pivot.to_numpy(), **ratio_options)
    records = [
        {"protein": protein, "sample": sample, "intensity": intensity}
        for sample, intensity in zip(samples, intensities)
        if intensity > 0
    ]
    return records, stabilized_pairs


def _prepare_input(peptide_df, columns, run_column):
    """Filter invalid rows and retain the supplied species and sample/run axes."""
    protein_column, peptide_column, intensity_column, sample_column = columns
    frame = peptide_df.copy()
    frame = frame[np.isfinite(frame[intensity_column]) & (frame[intensity_column] > 0)]
    required = [protein_column, peptide_column, sample_column]
    if run_column is not None:
        required.append(run_column)
    frame = frame.dropna(subset=required)
    if {"PeptideSequence", "PrecursorCharge"}.issubset(frame.columns):
        frame = frame.dropna(subset=["PeptideSequence", "PrecursorCharge"])
        frame["PrecursorCharge"] = pd.to_numeric(frame["PrecursorCharge"])
        charge = frame["PrecursorCharge"]
        frame = frame[np.isfinite(charge) & (charge > 0) & (charge % 1 == 0)].copy()
        frame["_maxlfq_species"] = list(
            zip(frame["PeptideSequence"], frame["PrecursorCharge"])
        )
        peptide_column = "_maxlfq_species"
    if run_column is not None:
        frame["_maxlfq_sample_run"] = list(zip(frame[sample_column], frame[run_column]))
        axis_column = "_maxlfq_sample_run"
    else:
        axis_column = sample_column
    return frame, peptide_column, axis_column


def _format_results(all_results, protein_column, sample_column, run_column):
    """Restore caller identifiers and optional run columns in positive output."""
    result = pd.DataFrame(
        [row for group, _ in all_results for row in group],
        columns=["protein", "sample", "intensity"],
    ).rename(
        columns={
            "protein": protein_column,
            "sample": sample_column,
            "intensity": "Intensity",
        }
    )
    if run_column is not None:
        pairs = result[sample_column].tolist()
        result[sample_column] = [pair[0] for pair in pairs]
        result[run_column] = [pair[1] for pair in pairs]
        result = result[[protein_column, sample_column, run_column, "Intensity"]]
    return result


@PluginRegistry.register("quantification", "maxlfq")
@dataclass
class MaxLFQQuantification(ProteinQuantificationMethod):
    """Independent MaxLFQ ratio solver for prepared linear intensities.

    ``min_ratio_count`` is the number of shared species required for EACH sample
    pair (default 2). ``min_peptides`` is its legacy argument alias. Proteins or
    samples without a valid ratio are not replaced by median aggregation.
    ``force_builtin`` is retained as a no-op for compatibility. MaxLFQ never
    dispatches to DirectLFQ. ``threads`` follows joblib's worker-count semantics.
    ``stabilize`` enables Eq. 5's overlap-dependent large-ratio stabilization.
    """

    min_peptides: int = 2
    threads: int = -1
    verbose: int = 0
    force_builtin: bool = False
    min_ratio_count: Optional[int] = field(default=None, kw_only=True)
    stabilize: bool = field(default=False, kw_only=True)
    using_directlfq: ClassVar[bool] = False

    def __post_init__(self):
        self.min_ratio_count = (
            self.min_peptides if self.min_ratio_count is None else self.min_ratio_count
        )
        _maxlfq_solve_protein(
            np.empty((0, 0)),
            min_ratio_count=self.min_ratio_count,
            stabilize=self.stabilize,
        )
        self.min_peptides = self.min_ratio_count

    @property
    def name(self) -> str:
        return "MaxLFQ"

    def quantify(
        self,
        peptide_df: pd.DataFrame,
        protein_column: str = PROTEIN_NAME,
        peptide_column: str = PEPTIDE_CANONICAL,
        intensity_column: str = NORM_INTENSITY,
        sample_column: str = SAMPLE_ID,
        run_column: Optional[str] = None,
    ) -> pd.DataFrame:
        """Quantify sample or (sample, run) profiles, preserving species identity.

        When PeptideSequence and PrecursorCharge are available, use their pair
        as the species identifier. Otherwise use the supplied peptide column;
        a canonical-only input cannot recover discarded charge/modification data.
        Missing and invalid intensities are excluded; empty outputs keep headers.
        """
        frame, peptide_column, axis_column = _prepare_input(
            peptide_df,
            (protein_column, peptide_column, intensity_column, sample_column),
            run_column,
        )
        samples = sorted(frame[axis_column].unique(), key=_identifier_sort_key)
        groups = frame.groupby(protein_column, observed=True)
        all_results = Parallel(n_jobs=self.threads, verbose=self.verbose)(
            delayed(_process_protein)(
                protein,
                group,
                (peptide_column, intensity_column, axis_column),
                samples,
                {
                    "min_ratio_count": self.min_ratio_count,
                    "stabilize": self.stabilize,
                },
            )
            for protein, group in groups
        )
        logger.info(
            "MaxLFQ min_ratio_count=%d stabilize=%s: %d protein groups and %d "
            "sample pairs have positive stabilization weights",
            self.min_ratio_count,
            self.stabilize,
            sum(count > 0 for _, count in all_results),
            sum(count for _, count in all_results),
        )
        return _format_results(all_results, protein_column, sample_column, run_column)
